sentry-ruby/spec/sentry/hub_spec.rb

```diff
@@ -4,71 +4,228 @@
   let(:configuration) do
     config = Sentry::Configuration.new
     config.dsn = DUMMY_DSN
+    config.transport.transport_class = Sentry::DummyTransport
     config
   end
   let(:client) { Sentry::Client.new(configuration) }
+  let(:transport) { client.transport }
   let(:scope) { Sentry::Scope.new }
 
   subject { described_class.new(client, scope) }
 
-  before do
-    allow(client).to receive(:send_event)
-  end
-
   describe '#capture_message' do
     let(:message) { "Test message" }
-    let(:options) { { tags: { foo: "bar" }, server_name: "foo.local" } }
 
     it "returns an Event instance" do
-      expect(subject.capture_message(message, options)).to be_a(Sentry::Event)
+      expect(subject.capture_message(message)).to be_a(Sentry::Event)
     end
 
     it 'initializes an Event, and sends it via the Client' do
-      expect(client).to receive(:send_event)
+      expect do
+        subject.capture_message(message)
+      end.to change { transport.events.count }.by(1)
+    end
+
+    context "with custom attributes" do
+      it "updates the event with custom attributes" do
+        subject.capture_message(message, tags: { foo: "bar" })
+
+        event = transport.events.last
+        expect(event.tags).to eq({ foo: "bar" })
+      end
 
-      subject.capture_message(message, options)
+      it "merges the contexts/tags/extrac with what the scope already has" do
+        scope.set_tags(old_tag: true)
+        scope.set_contexts(old_context: true)
+        scope.set_extras(old_extra: true)
+
+        subject.capture_message(
+          message,
+          tags: { new_tag: true },
+          contexts: { new_context: true },
+          extra: { new_extra: true }
+        )
+
+        event = transport.events.last
+        expect(event.tags).to eq({ new_tag: true, old_tag: true })
+        expect(event.contexts).to eq({ new_context: true, old_context: true })
+        expect(event.extra).to eq({ new_extra: true, old_extra: true })
+
+        expect(scope.tags).to eq(old_tag: true)
+        expect(scope.contexts).to eq(old_context: true)
+        expect(scope.extra).to eq(old_extra: true)
+      end
     end
 
-    it 'yields the event to a passed block' do
-      expect(client).to receive(:send_event)
+    context "with custom scope" do
+      let(:new_scope) do
+        scope = Sentry::Scope.new
+        scope.set_tags({ custom_scope: true })
+        scope
+      end
+
+      it "accepts a custom scope" do
+        subject.capture_message(message, scope: new_scope)
+
+        event = transport.events.last
+        expect(event.tags).to eq({ custom_scope: true })
+      end
+    end
```

Comment:
Can we also set something here on the scope to make sure the passed scope instance is used and not the global one?

Reply:
this is exactly this test for: only the `new_scope` has the `custom_scope: true` tag and the global one doesn't